src/s390x/dasd.rs

```diff
@@ -0,0 +1,284 @@
+
+        count -= 1;
+        if count != 0 {
+            partitions.push(format!("[{}, {}, native]", start, end));
+        } else {
+            partitions.push(format!("[{}, last, native]", start));
+        };
+        start = end + 1;
+    }
+    Ok((dd_options, partitions))
+}
+
+/// Get disk bus id
+///
+/// # Arguments
+/// * `dasd` - dasd device, i.e. smth like /dev/dasda
+fn bus_id(dasd: &str) -> Result<String> {
+    let cmd = Command::new("lszdev")
+        .arg("-n")
+        .arg("-c")
+        .arg("ID")
+        .arg("--by-node")
+        .arg(dasd)
+        .output()
+        .chain_err(|| "failed to execute lszdev")?;
+    let s = cmd
+        .stdout
+        .iter()
+        .map(|&c| c as char)
+        .filter(|&c| c != '\n' && c != '\r')
+        .collect::<String>();
+    eprintln!("Bus ID: {}", s);
+    Ok(s)
+}
+
+/// Check if disk is already formatted or not
+///
+/// # Arguments
+/// * `dasd` - dasd device, i.e. smth like /dev/dasda
+fn is_unformatted(dasd: &str) -> Result<bool> {
+    let id = bus_id(dasd)?;
+    let path = format!("/sys/bus/ccw/devices/{}/status", id);
+    let contents = read_to_string(&path).chain_err(|| format!("reading {}", path))?;
+    Ok(contents.contains("unformatted"))
+}
+
+/// Check if disk is valid or not
+///
+/// # Arguments
+/// * `dasd` - dasd device, i.e. smth like /dev/dasda
+fn is_invalid(dasd: &str) -> Result<bool> {
+    let cmd = Command::new("fdasd").arg("-p").arg(dasd).output()?;
+    let output = cmd.stdout.iter().map(|&c| c as char).collect::<String>();
+    Ok(output.contains("disk label block is invalid"))
+}
+
+/// Perform low-level format. This step is necessary before any further disk usage
+///
+/// # Arguments
+/// * `dasd` - dasd device, i.e. smth like /dev/dasda
+fn low_level_format(dasd: &str) -> Result<()> {
+    if !is_unformatted(dasd)? {
+        eprintln!("Skipping low-level format for: {}", dasd);
+        return Ok(());
+    }
+    eprintln!("Performing low-level format for: {}", dasd);
+    let status = Command::new("dasdfmt")
+        .arg("--blocksize")
+        .arg("4096")
+        .arg("--disk_layout")
+        .arg("cdl")
+        .arg("--mode")
+        .arg("full")
+        .arg("-y")
+        .arg("-p")
+        .arg("-v")
+        .arg(dasd)
+        .stdout(std::process::Stdio::inherit())
+        .status()
+        .chain_err(|| "failed to execute dasdfmt")?;
+    check(status).chain_err(|| "low-level format")
+}
+
+/// Format disk and create partitions
+///
+/// # Arguments
+/// * `dasd` - dasd device, i.e. smth like /dev/dasda
+/// * `partitions` - configuration strings
+fn make_partitions(dasd: &str, partitions: &[String]) -> Result<()> {
+    let config_filename = "/tmp/fdasd.conf";
+
+    write(config_filename, partitions.join("\n"))
+        .chain_err(|| format!("writing {}", config_filename))?;
+
+    if try_format(dasd, &config_filename).is_err() {
+        default_format(dasd)?;
+        try_format(dasd, &config_filename)?;
```

Comment:
Is an auto-format followed by a config-based format likely to actually fix the problem?

Reply:
unfortunately now. if disk has no valid table, than it should be auto-formatted

Reply:
sorry, misunderstood the question. yes, auto-format fixes all issues, and than config-based format  works fine. (it may fail only if config itslef is not valid )